### ISSUE TITLE
As a developer, I want to know the sequence of API calls needed to upload a new collection

### ISSUE THREAD

**Issue by r-b-g-b** [openoakland/woeip#154]
@nbilenko and I have started discussing the sequence of API calls that the front end developers need to implement in order to perform a complete upload of a new data collection. This is just an attempt to write it all down and start hashing out the details. Here is one version of how it can go. We don't need to worry too much about what is supported by the API currently -- even this version will require some minor tweaks to the API.

1. **(UI)** User drags and drops two files, a GPS log and Dustrak air quality log.
2. **(Frontend)** 
 - Parses GPS and AQ data files
 - Validates data fields

(This POST request might need to be changed to reflect more front end data parsing)
```
http://api.lvh.me/collection POST
{
    "upload_files": [
        "starts_at": "2019-01-01T12:00:00", "ends_at": "2019-01-01T13:00",
        {"file_name": "gps.log", "file_data": gps_data},
        {"file_name": "dustrak.log", "file_data": aq_data}
    ]
}
```

3. **(Backend)**
 - Creates a `Collection` instance and two `CollectionFile` instances. Also create any new `Pollutant`/`Device`, etc. as needed.
 - For each new collection file, sets the `collection_file.collection` to the new `collection`.
 - Returns entire collection data (including nested collection files, sensors)
```
{
    "starts_at": "2020-01-01T12:00:00Z",
    "ends_at": "2020-01-01T12:00:00Z",
    "collection_files": [
        {
            "collection": 44,
            "sensor": {
                "name": "GPS",
                "device": "http://api.lvh.me/device/1",
                "pollutant": "http://api.lvh.me/pollutant/8",
                "unit": "lat/lon"
            },
            "user": 1,
            "file": "http://api.lvh.me/media/data/dustrak_HK9u4VW.log",
            "uploaded_at": "2019-11-13T06:05:50.406805Z",
            "processor_version": "1",
            "processed_at": "2019-11-14T05:33:50Z"},

        {
            "collection": 44,
            "sensor": {
                "name": "Dustrak",
                "device": "http://api.lvh.me/device/1",
                "pollutant": "http://api.lvh.me/pollutant/8",
                "unit": "ppm"
            },
            "user": 1,
            "file": "http://api.lvh.me/media/data/gps_FhlgWM4.log",
            "uploaded_at": "2019-11-13T06:05:50.400169Z",
            "processor_version": "1",
            "processed_at": "2019-11-14T05:34:06Z"
        }
    ]
}
```
4. **(Frontend)** Renders form for user data validation
5. **(UI)** User validates data
6. **(Frontend)** API request to update data

```
http://api.lvh.me/collection PATCH

{
...
}
```

7. **(Backend)** Updates collection, collection files, and sensor info. Commits them to the database.
8. **(Backend)** Joins files and saves `PollutantValue` for each data point.

### Related

- Issue #132 also surfaces potential API call requirements, so let's make sure we're reviewing these together. The latest wires were created with #132 in mind: [WOAQ-Upload-v5-Flow](https://user-images.githubusercontent.com/32755410/72934294-31121700-3d18-11ea-80dc-26e622595b95.png)
- [WOAQ-Data-Flow-Diagram-v5.pdf](https://github.com/openoakland/woeip/files/4157171/WOAQ-Data-Flow-Diagram-v5.pdf)

**Comment by theecrit** [comment]
Related: https://github.com/openoakland/woeip/issues/132

**Comment by theecrit** [comment]
~Meeting is set for **Thursday, Tues. Feb 6, 2020** from 5:30-6:30pm:~

Rescheduling to accommodate more folks.

**Comment by jayqi** [comment]
So per the changes from #181, there is only a single API call to do all of the necessary entity creation and processing in the backend. 

What I think is perhaps a simpler flow vs. the top comment: Steps 1-3 send the files to the backend, then Step 8 does the join. I think that is enough to get us through development of #129.

Steps 4-7 can be a separate thing after Step 8 is already done.
